pkg/referrerstore/oras/authprovider/azure/azureidentity.go

```diff
@@ -0,0 +1,173 @@
+}
+
+var (
+	// ErrorNoAuth indicates that no credentials are provided.
+	ErrorNoAuth    = fmt.Errorf("no credentials provided for Azure cloud provider")
+	ErrorNilEnv    = fmt.Errorf("env is nil pointer for GetServicePrincipalToken")
+	ErrorNilConfig = fmt.Errorf("config is nil pointer for GetServicePrincipalToken")
+)
+
+const (
+	azureManagedIdentityAuthProviderName string = "azureManagedIdentity"
+)
+
+// init calls Register for our Azure Workload Identity provider
+func init() {
+	provider.Register(azureManagedIdentityAuthProviderName, &azureManagedIdentityProviderFactory{})
+}
+
+// Create returns an azureManagedIdentityAuthProvider
+func (s *azureManagedIdentityProviderFactory) Create(authProviderConfig provider.AuthProviderConfig) (provider.AuthProvider, error) {
+	conf := azureManagedIdentityAuthProviderConf{}
+	authProviderConfigBytes, err := json.Marshal(authProviderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal(authProviderConfigBytes, &conf); err != nil {
+		return nil, fmt.Errorf("failed to parse auth provider configuration: %v", err)
+	}
+
+	tenant := os.Getenv("AZURE_TENANT_ID")
+	if tenant == "" {
+		return nil, fmt.Errorf("AZURE_TENANT_ID environment variable is empty")
+	}
+	client := os.Getenv("AZURE_CLIENT_ID")
+	if client == "" {
+		client = conf.ClientID
+		if client == "" {
+			return nil, fmt.Errorf("AZURE_CLIENT_ID environment variable is empty")
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+	// retrieve an AAD Access token
+	token, err := getManagedIdentityToken(context.Background(), client)
+	if err != nil {
+		return nil, err
+	}
+
+	return &azureManagedIdentityAuthProvider{
+		identityToken: token,
+		clientID:      client,
+		tenantID:      tenant,
+	}, nil
+}
+
+// Enabled checks for non empty tenant ID and AAD access token
+func (d *azureManagedIdentityAuthProvider) Enabled(ctx context.Context) bool {
+	if d.clientID == "" {
+		return false
+	}
+
+	if d.tenantID == "" {
+		return false
+	}
+
+	if d.identityToken.Token == "" {
+		return false
+	}
+
+	return true
+}
+
+// Provide returns the credentials for a specified artifact.
+// Uses Managed Identity to retrieve an AAD access token which can be
+// exchanged for a valid ACR refresh token for login.
+func (d *azureManagedIdentityAuthProvider) Provide(ctx context.Context, artifact string) (provider.AuthConfig, error) {
+	if !d.Enabled(ctx) {
+		return provider.AuthConfig{}, fmt.Errorf("azure managed identity provider is not properly enabled")
+	}
+	// parse the artifact reference string to extract the registry host name
+	artifactHostName, err := provider.GetRegistryHostName(artifact)
+	if err != nil {
+		return provider.AuthConfig{}, err
+	}
+
+	// need to refresh AAD token if it's expired
+	if time.Now().Add(time.Minute * 5).After(d.identityToken.ExpiresOn) {
+		newToken, err := getManagedIdentityToken(ctx, d.clientID)
+		if err != nil {
+			return provider.AuthConfig{}, errors.Wrap(err, "could not refresh AAD token")
```

Comment:
best to include auth provider name in error msg

Reply:
added